reputation/reputation.go

```diff
@@ -201,16 +210,114 @@ func ModifyRep(ctx context.Context, conf *models.ReputationConfig, guildID int64
 	return
 }
 
-func insertUpdateUserRep(ctx context.Context, guildID, userID int64, amount int64) (err error) {
+const (
+	MaxRepRoles        = 5
+	MaxRepRolesPremium = 25
+)
+
+func GuildMaxRepRoles(guildID int64) int {
+	if isPrem, _ := premium.IsGuildPremium(guildID); isPrem {
+		return MaxRepRolesPremium
+	}
+	return MaxRepRoles
+}
+
+func UpdateRepRoles(gs *dstate.GuildSet, ms *dstate.MemberState, rep int64) error {
+	repRoles, err := models.ReputationRoles(models.ReputationRoleWhere.GuildID.EQ(gs.ID)).AllG(context.Background())
+	if err != nil {
+		return err
+	}
+	if len(repRoles) == 0 {
+		return nil // nothing to do
+	}
+
+	botMember, err := bot.GetMember(gs.ID, common.BotUser.ID)
+	if err != nil {
+		return err
+	}
+	botHighestRole := bot.MemberHighestRole(gs, botMember)
+
+	add, remove := computeRoleChanges(ms.Member.Roles, rep, repRoles)
+	add, remove = filterActionable(add, gs, botHighestRole), filterActionable(remove, gs, botHighestRole)
+	if len(add) == 0 && len(remove) == 0 {
+		return nil // nothing to do
+	}
+
+	// Can we edit the member's roles in one go?
+	if common.IsRoleAbove(botHighestRole, bot.MemberHighestRole(gs, ms)) {
+		return bulkEditRoles(ms, add, remove)
+	}
+
+	// Otherwise, we need to add and remove roles one by one to avoid a permissions error.
+	// Issue the API requests sequentially to help avoid ratelimits.
+	for _, r := range add {
+		if err = common.BotSession.GuildMemberRoleAdd(gs.ID, ms.User.ID, r); err != nil {
+			return err
+		}
+	}
+	for _, r := range remove {
+		if err := common.BotSession.GuildMemberRoleRemove(gs.ID, ms.User.ID, r); err != nil {
+			return err
+		}
+	}
```

Comment:
I don't really like the loop here, but as above I don't know if there's a way to avoid it. There is some potential for abuse here, I suppose, but I think it is balanced by the fact there's only one role to add/remove in the common case. (The worst case happens when a premium server has set up the full 25 rep roles, a member above the bot has all the rep roles, and someone resets their reputation in which case we will issue 25 API requests.) One can already write dubious loops like this using custom commands as well.